### ISSUE TITLE
Make typing annotations more flexible

### ISSUE THREAD

**Issue by pvdrz** [pvdrz/pijama_legacy#92]
After https://github.com/christianpoveda/pijama/pull/91 lands. Our type system will be able to infer the type of some functions even if they don't have all their type annotations.

Right now, the following program is allowed
```elixir
fn add(x: Int, y: Int): Int do
  x + y
end
```
It is also OK to omit the return type annotation
```elixir
fn add(x: Int, y: Int) do
  x + y
end
```
However we cannot omit the annotations for `x` and `y`. Thus, this program won't be accepted by the compiler
```elixir
fn add(x, y) do
  x + y
end
```
Ironically, if we write an anonymous function. We cannot annotate its return type, so this program will also be rejected
```elixir
fn (x: Int, y: Int): Int do
  x + y
end
```
Even before  https://github.com/christianpoveda/pijama/pull/91, our type system is more than capable of handling some of those cases. However those programs are rejected, not only because the parser doesn't allow it, but because there is no way to represent the presence (or lack of) a type annotation over `pijama::mir::Term` in a general way.

A first step to solve all these issues would be to add a new `Ascription` or `Cast` variant to `pijama::mir::Term` which looks like this:
```rust
enum Term<'a> {
    Cast(Box<Located<Term<'a>>, Located<Ty<'a>>)
}
```
Which has a similar role as `pijama::ty::Binding`, i.e. `Term::Cast(term, ty)` declares that `term` must have type `ty`. The steps to add this new `Term::Cast` variant would be

- [ ] Add it to the `mir::Term` enum.
- [ ] Add a suitable `Display` implementation, maybe `Term::Cast(term, ty)` could be displayed as `(term as ty)`.
- [ ] Add a new `type_of_cast` inside `pijama::ty::ty_check::Context` ensuring that the type of the term in the cast matches the type stated in the cast.
- [ ] Add a new branch to `pijama::lir::lower::Context::remove_names` with the pattern `MirTerm::Cast(term, _)` returning  `self.remove_names(term)`.

Fixing this issue is a tiny tour over the codebase and there are more details involved in these steps. If you want to tackle this issue let me know and I'll assist you however I can :)

Update: It seems this is not as useful as I thought it would be. We have to do this eventually but to solve other issues that we don't have yet.

**Comment by DarkDrek** [comment]
I would like to do that probably after #91 lands.

**Comment by pvdrz** [comment]
Sure! but #91 will land after @seanchen1991 migrates `ast` to a new crate because I want to maintain a difference between the types the used annotated an the types used in #91 for inference.

**Comment by DarkDrek** [comment]
That won't be a problem I can wait 😴. 
Do you think that `Term::Cast` may conflict later when we add some kind of casting between types? Or will these just user level functions?
```
fn as_bool(x: Int): Bool do
    if x == 0 do
        false
    else
        true
    end
end
```

**Comment by pvdrz** [comment]
I think we can handle that interaction without much trouble. That being said, I'd rather have those casts as user level functions like `std::convert::Into` in Rust.

**Comment by pvdrz** [comment]
@DarkDrek  #91 got merged so we're ready to start with this. :)

Update: After discussing this with my pillow I think this might take us to the wrong path. I'd suggest we keep this one on hold.
